ops/testing.py

```diff
@@ -2586,6 +2606,118 @@ class NonAbsolutePathError(Exception):
+        """
+
+    @abstractmethod
+    def list_files(self, path: StrOrPath, *, pattern: Optional[str] = None,
+                   itself: bool = False) -> List['FileInfo']:
+        """Return list of directory entries from given path on remote system.
+
+        See :meth:`model.Container.list_files` for details.
+        """
+
+    @abstractmethod
+    def push_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath):
+        """Recursively push a local path or files to the remote system.
+
+        See :meth:`model.Container.push_path` for details.
+        """
+
+    @abstractmethod
+    def pull_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath):
+        """Recursively pull a remote path or files to the local system.
+
+        See :meth:`model.Container.pull_path` for details.
+        """
+
+    @abstractmethod
+    def exists(self, path: str) -> bool:
+        """Return true if the path exists on the container filesystem.
+
+        See :meth:`model.Container.exists` for details.
+        """
+
+    @abstractmethod
+    def isdir(self, path: str) -> bool:
+        """Return true if a directory exists at the given path on the container filesystem.
+
+        See :meth:`model.Container.isdir` for details.
+        """
+
+    @abstractmethod
+    def make_dir(
+            self, path: str, *, make_parents: bool = False, permissions: Optional[int] = None,
+            user_id: Optional[int] = None, user: Optional[str] = None,
+            group_id: Optional[int] = None, group: Optional[str] = None):
+        """Create a directory on the remote system with the given attributes.
+
+        See :meth:`model.Container.make_dir` for details.
+        """
+
+    @abstractmethod
+    def remove_path(self, path: str, *, recursive: bool = False):
+        """Remove a file or directory on the remote system.
+
+        See :meth:`model.Container.remove_path` for details.
+        """
+
+
+class _TestingContainerFilesystemProxy:
+    """A proxy providing unrestricted filesystem access for a specified container."""
+
+    def __init__(self, container: model.Container, backend: _TestingPebbleClient):
+        self._container = container
+        self._backend = backend
+
+    def __getattr__(self, name: str):
+        if name not in ContainerFilesystem.__dict__ or name.startswith("_"):
+            raise AttributeError(f'object has no attribute {name!r}')
+        func = getattr(self._container, name)
+
+        @functools.wraps(func)
+        def proxy_function(*args: Any, **kwargs: Any) -> Any:
+            can_connect = self._backend._can_connect(self._container._pebble)
+            self._backend._set_can_connect(self._container._pebble, True)
+            result = func(*args, **kwargs)
+            self._backend._set_can_connect(self._container._pebble, can_connect)
```

Comment:
Let's do a `try ... finally` here just in case the test is doing some exception handling, to avoid `can_connect` remaining in an unexpected state. I think that would look like this:

```suggestion
            try:
                result = func(*args, **kwargs)
            finally:
                self._backend._set_can_connect(self._container._pebble, can_connect)
```

Reply:
Oops, that's a big one, thank you for spotting this, fixed! I also added a test case for the exception safety.